src/Compilers/CSharp/Portable/Symbols/TypeSymbolWithAnnotations.cs

```diff
@@ -639,7 +606,12 @@ public bool Equals(TypeSymbolWithAnnotations other, TypeCompareKind comparison)
                 return true;
             }
 
-            if (other.IsNull || !TypeSymbolEquals(other, comparison))
+            if (!HasType)
+            {
+                return !other.HasType;
```

Comment:
Don't we need to check the annotation here as well? Or at least assert that the annotations are the same?